README.txt
This is a CSE 491/Web dev project; see http://msu-web-dev.readthedocs.org/.

1. Running server.py on arctic.
===============================

In a subdirectory on arctic.cse.msu.edu, create a Python `virtualenv
<http://www.virtualenv.org/en/latest/>`__ using Python 2.7::

   python2.7 -m virtualenv cse491.env

Now, activate this virtualenv::

   source cse491.env/bin/activate.csh

(From now on, you'll want to make sure to have this virtualenv active
whenever you're working on server.py.)

Next, go to your cse491-serverz directory and run server.py::

   python server.py

Use CTRL-C to quit the server.

   push:  git push origin master
   swap branches: git checkout branchname  (-b for new branch)
   push into github: git push origin master:other
       where master is the master branch and other is the name of hithub branch
         - these can match, or not.  easier if you keep them consistent tho



ChangeLog
2014-01-21   Ryan Miller
 * Added some response text Hello, world!

 * Initialized repository, created server.py and README.

 * hw2, added some unit tests

2014-01-28
 * fixed HW2 - added ability handle POST requests
   and added test case for POSTs

2014-01-30
 * Added form submissions for both GET/POST
 * Refactored main()
 * Fixed/added test cases for the form

 2014-02-06
 * Added 404 page
 * Added a few template pages
 * Adjusted c.recv to accept any data size
 * Added test cases for new content


test_server.py
import server


class FakeConnection(object):
    """
    A fake connection class that mimics a real TCP socket for the purpose
    of testing socket I/O.
    """

    def __init__(self, to_recv):
        self.to_recv = to_recv
        self.sent = ""
        self.is_closed = False

    def recv(self, n):
        if n > len(self.to_recv):
            r = self.to_recv
            self.to_recv = ""
            return r

        r, self.to_recv = self.to_recv[:n], self.to_recv[n:]
        return r

    def send(self, s):
        self.sent += s

    def close(self):
        self.is_closed = True

# Test a basic GET call.


def test_handle_connection():
    conn = FakeConnection("GET / HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n' + 'Content-type: text/html\r\n'
                       '\r\n'
                       '<body>'
                       '<h1>Hello, world.</h1>'
                       '<p>'
                       'This is Ryan Miller\'s (mill1256) Web server.</br>'
                       '&nbsp&nbsp<a href=/content>Content</a></br>'
                       '&nbsp&nbsp<a href=/file>File</a></br>'
                       '&nbsp&nbsp<a href=/image>Image</a></br>'
                       '&nbsp&nbsp<a href=/form>Form</a></br>'
                       '</p>'
                       '</body>')
     
    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_content_connection():
    conn = FakeConnection("GET /content HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n'
                       'Content-type: text/html\r\n'
                       '\r\n'
                       '<h1>Content path!</h1>')

    server.handle_connection(conn)
    assert expected_return in conn.sent, 'Got: %s' % (repr(conn.sent),)


def test_handle_file_connection():
    conn = FakeConnection("GET /file HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n'
                       'Content-type: text/html\r\n'
                       '\r\n'
                       '<h1>file path!</h1>')
    server.handle_connection(conn)
    assert expected_return in conn.sent, 'Got: %s' % (repr(conn.sent),)


def test_handle_image_connection():
    conn = FakeConnection("GET /image HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n'
                       'Content-type: text/html\r\n'
                       '\r\n'
                       '<h1>image path!</h1>')

    server.handle_connection(conn)
    assert expected_return in conn.sent, 'Got: %s' % (repr(conn.sent),)


def test_handle_post_request():
    conn = FakeConnection("POST / HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n'
                       'Content-type: text/html\r\n'
                       '\r\n'
                       '<h1>Hello, world!</h1>')

    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_form_connection():
    conn = FakeConnection("GET /form HTTP/1.0\r\n\r\n")
    expected_return = ("HTTP/1.0 200 OK\r\n"
                       "Content-type: text/html\r\n"
                       "\r\n"
                       "<form action='/submit' method='GET'>"
                       "First Name: <input type='text' name='firstname'></br>"
                       "Last Name: <input type='text' name='lastname'></br>"
                       "<button name='submitButton'"
                       " type='submit'>Submit</button>"
                       "</form>")

    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_form_valid_submission():
    conn = FakeConnection("GET /submit?firstname=Ryan&lastname=Miller&submitButton= HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n' +
                       'Content-type: text/html\r\n' +
                       '\r\n' +
                       '<h1>Hi Ryan Miller</h1>')

    server.handle_connection(conn)
    print expected_return
    print repr(conn.sent)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_form_post():
    conn = FakeConnection("POST /form/submit? HTTP/1.0"
                          "&firstname=Ryan&lastname=Miller\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n' +
                       'Content-type: text/html\r\n' +
                       '\r\n' +
                       '<h1>Hi Ryan Miller\r\n\r\n</h1>')

    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_form_post_invalid():
    conn = FakeConnection("POST /form/submit? HTTP/1.0"
                          "&firstname=&lastname=\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n' +
                       'Content-type: text/html\r\n' +
                       '\r\n' +
                       '<h1>Warning, first and last name must be entered</h1>')

    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)


def test_handle_form_invalid_submission():
    conn = FakeConnection("GET /submit?firstname=&lastname=&submitButton= HTTP/1.0\r\n\r\n")
    expected_return = ('HTTP/1.0 200 OK\r\n'
                       'Content-type: text/html\r\n'
                       '\r\n'
                       '<h1>Warning, first and last name must be entered</h1>')

    server.handle_connection(conn)
    assert conn.sent == expected_return, 'Got: %s' % (repr(conn.sent),)

    
def test_handle_form_invalid_submission():
    conn = FakeConnection("GET /iAmAGoon HTTP/1.0\r\n\r\n")
    expected_return = ('src="http://www.finerminds.com')

    server.handle_connection(conn)
    assert expected_return in conn.sent, 'Got: %s' % (repr(conn.sent),)


server.py
#!/usr/bin/env python
import random
import socket
import urlparse
import cgi
import urllib2
import render

RSP_SUCC_HDR = 'HTTP/1.0 200 OK\r\nContent-type: text/html\r\n\r\n'

def main():
    s = socket.socket()         # Create a socket object
    host = socket.getfqdn()     # Get local machine name
    port = random.randint(8000, 9999)
    s.bind((host, port))        # Bind to the port

    print 'Starting server on', host, port
    print 'The Web server URL for this would be http://%s:%d/' % (host, port)

    s.listen(5)                 # Now wait for client connection.

    print 'Entering infinite loop; hit CTRL-C to exit'
    while True:
        # Establish connection with client.
        c, (client_host, client_port) = s.accept()
        handle_connection(c)
        print 'Got connection from', client_host, client_port


def generate_content_page(*args):
        conn = args[0]
        file_object = open('templates/contentPage.html')
        html_content = file_object.read()
        file_object.close()
        conn.send(RSP_SUCC_HDR)
        conn.send(html_content)


def generate_file_page(*args):
        conn = args[0]
        file_object = open('templates/filePage.html')
        html_content = file_object.read()
        file_object.close()
        conn.send(RSP_SUCC_HDR)
        conn.send(html_content)


def generate_image_page(*args):
        conn = args[0]
        file_object = open('templates/imagePage.html')
        html_content = file_object.read()
        file_object.close()
        conn.send(RSP_SUCC_HDR)
        conn.send(html_content)


def generate_form_page(*args):
        conn = args[0]
        conn.send("HTTP/1.0 200 OK\r\n")
        conn.send("Content-type: text/html\r\n")
        conn.send("\r\n")
        conn.send("<form action='/submit' method='GET'>")
        conn.send("First Name: <input type='text' name='firstname'></br>")
        conn.send("Last Name: <input type='text' name='lastname'></br>")
        conn.send("<button name='submitButton' type='submit'>Submit</button>")
        conn.send("</form>")


def generate_form_submission_page(*args):
        conn = args[0]
        headers = {}
        socket_data = args[1]
        fp = urllib2.StringIO(socket_data)
        parsed_url = urlparse.urlparse(socket_data)
        trimmed_url_list = parsed_url.query.split("\r\n\r\n")
        header_list = trimmed_url_list[0].split("\r\n")
        print header_list
        for entry in header_list:
            try:
                entry = entry.split(' ')
                headers[entry[0]] = entry[1]
            except:
                print 'Error parsing header dictionary. url:', parsed_url
        field_storage = cgi.FieldStorage(fp, headers)
        print socket_data
        if 'enctype:' in field_storage.headers.keys():
            if field_storage.headers['enctype:'] == 'multipart/form-data':
                print 'multiformdata!'
        else:
            try:
                #protect against empty inputs
                first_name = urlparse.parse_qs(parsed_url.query)['firstname'][0]
                last_name = urlparse.parse_qs(parsed_url.query)['lastname'][0]
                conn.send('HTTP/1.0 200 OK\r\n')
                conn.send('Content-type: text/html\r\n')
                conn.send('\r\n')
                conn.send('<h1>Hi %s %s</h1>' % (first_name, last_name))
            except:
                #if except block is hit, one or more of the fields were empty
                conn.send('HTTP/1.0 200 OK\r\n')
                conn.send('Content-type: text/html\r\n')
                conn.send('\r\n')
                conn.send('<h1>Warning, first and last name must be entered</h1>')
                #@question - why does following link not direct to the home
                #page on the first attempt, but does on the 2nd attempt?
                #conn.send('<a href=/>Home</a>')


def generate_home_page(*args):
        conn = args[0]
        conn.send('HTTP/1.0 200 OK\r\n' + 'Content-type: text/html\r\n')
        conn.send('\r\n')
        conn.send('<body>')
        conn.send('<h1>Hello, world.</h1>')
        conn.send('<p>')
        conn.send('This is Ryan Miller\'s (mill1256) Web server.</br>')
        conn.send('&nbsp&nbsp<a href=/content>Content</a></br>')
        conn.send('&nbsp&nbsp<a href=/file>File</a></br>')
        conn.send('&nbsp&nbsp<a href=/image>Image</a></br>')
        conn.send('&nbsp&nbsp<a href=/form>Form</a></br>')
        conn.send('</p>''</body>')


def generate_post_page(*args):
        conn = args[0]
        conn.send('HTTP/1.0 200 OK\r\n')
        conn.send('Content-type: text/html\r\n')
        conn.send('\r\n')
        conn.send('<h1>Hello, world!</h1>')


def generate_post_form_page(*args):
        socket_data = args[1]
        parsed_url = urlparse.urlparse(socket_data)
        conn = args[0]
        try:
            #protect against empty inputs
            first_name = urlparse.parse_qs(parsed_url.query)['firstname'][0]
            last_name = urlparse.parse_qs(parsed_url.query)['lastname'][0]
            conn.send('HTTP/1.0 200 OK\r\n')
            conn.send('Content-type: text/html\r\n')
            conn.send('\r\n')
            conn.send('<h1>Hi %s %s</h1>' % (first_name, last_name))
        except:
            #if except block is hit, one or more of the fields were empty
            conn.send('HTTP/1.0 200 OK\r\n')
            conn.send('Content-type: text/html\r\n')
            conn.send('\r\n')
            conn.send('<h1>Warning, first and last name must be entered</h1>')


def generate_404_page(*args):
        conn = args[0]
        conn.send(RSP_SUCC_HDR)
        # image credit to finerminds.com
        conn.send('<body><img ')
        conn.send('src="http://www.finerminds.com/wp-content')
        conn.send('/blogs.dir/33/files/2011/01/404-cat.png">')
        conn.send('</body>')

def handle_connection(conn):
    sentinel_value = '\r\n\r\n'
    current_string = ''
    socket_data = ''
    while sentinel_value != current_string:
        current_char = conn.recv(1)
        if current_char == '\r' or current_char == '\n':
            socket_data += current_char
            current_string += current_char
        else:
            socket_data += current_char
            current_string = ''

    if 'POST /form' in socket_data:
        generate_post_form_page(conn, socket_data)
    elif 'POST /' in socket_data:
        generate_post_page(conn)
    elif "GET /content" in socket_data:
        generate_content_page(conn)
    elif "GET /file" in socket_data:
        generate_file_page(conn)
    elif "GET /image" in socket_data:
        generate_image_page(conn)
    elif 'GET /form' in socket_data:
        generate_form_page(conn)
    elif '/submit?' in socket_data:
        generate_form_submission_page(conn, socket_data)
    elif 'GET / ' in socket_data:
        generate_home_page(conn, socket_data)
    else:
        generate_404_page(conn)
    conn.close()


if __name__ == '__main__':
    main()



